### PR TITLE
[thanos] Add persistentVolumeClaim in rule component

### PR DESCRIPTION
**Pull request by JohnWillker**
| Q               | A
| --------------- | ---
| Bug fix?        | no|yes
| New feature?    | yes|no
| API breaks?     | no|yes
| Deprecations?   | no|yes
| Related tickets | no|yes
| License         | Apache 2.0

### What's in this PR?

Using Ruler for global rules evaluation in an [HA](https://thanos.io/tip/components/rule.md/#ruler-ha) mode, we have some problems when the rule pod dies, like lost all last 2h. To fix this problem we need to use a PersistentVolumeClain just like in Compact or in Prometheus in case of pod restart we maintain the data until the next upload to Bucket.

### Why?

Persist Ruler data for an HA use case.(Don't' lost data in ruler restarts)

### Checklist

- [x] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [x] User guide and development docs updated (if needed)
- [x] Related Helm chart(s) updated (if needed)

### CODE REVIEW

File: thanos/templates/rule-statefulset.yaml [hunk 1]
```diff
@@ -118,8 +118,10 @@ spec:
           readOnly: true
         {{- end }}
       volumes:
+      {{- if not .Values.rule.dataVolume.backend }}
```
Comment (tarokkk) [inline thanos/templates/rule-statefulset.yaml:121]:
```suggestion
      {{- if not .Values.rule.persistentVolumeClaim }}
```

File: thanos/README.md [hunk 2]
```diff
@@ -285,6 +285,8 @@ timePartioning:
 | rule.alertLabelDrop | Labels by name to drop before sending to alertmanager. This allows alert to be deduplicated on replica label (repeated). Similar Prometheus alert relabelling | [] |
 | rule.ruleOverrideName | Override rules file with custom configmap | "" |
 | rule.ruleFiles | See example in values.yaml | {}" |
+| rule.dataVolume.backend    | Data volume for the rule to store temporary data defaults to emptyDir.      | {} |
```
Comment (tarokkk) [inline thanos/README.md:288]:
```suggestion
```

File: thanos/values.yaml [hunk 3]
```diff
@@ -687,6 +687,24 @@ rule:
   replicaCount: 1
   # Kubernetes update strategy object as documented in https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies
   updateStrategy: {}
+  # Data volume for the rule to store temporary data defaults to emptyDir
```
Comment (tarokkk) [inline thanos/values.yaml:690]: Please remove this dataVolume part

**Review by tarokkk — CHANGES_REQUESTED** [review]
Thank you for your contribution. I simplified it a little bit.

File: thanos/values.yaml [hunk 4]
```diff
@@ -687,6 +687,24 @@ rule:
...
+  dataVolume:
+    backend: {}
+    #  persistentVolumeClaim:
+    #    claimName: rule-data-volume
```
Comment (JohnWillker) [inline thanos/values.yaml:694]:
```suggestion
```

File: thanos/templates/rule-statefulset.yaml [hunk 5]
```diff
@@ -161,4 +163,18 @@ spec:
       {{- with  .Values.rule.serviceAccount }}
       serviceAccountName: "{{ . }}"
       {{- end }}
+  {{- if .Values.rule.persistentVolumeClaim }}
+  volumeClaimTemplates:
+  - metadata:
+      name: {{ .Values.rule.dataVolume.backend.persistentVolumeClaim.claimName }}
```
Comment (JohnWillker) [inline thanos/templates/rule-statefulset.yaml:169]:
```suggestion
      name: {{ .Values.rule.persistentVolumeClaim.name }}
```

**Review by tarokkk — APPROVED** [review]

**Comment by JohnWillker** [comment]
@tarokkk Merge?

**Review by ahma — APPROVED** [review]
Thanks @JohnWillker LGTM

**Pull request merged**